src/ucp/core/ucp_rkey.c

```diff
@@ -213,61 +218,497 @@ UCS_PROFILE_FUNC(ssize_t, ucp_rkey_pack_memh,
+ *                                  memh_info_size_part1 is set to 0]
+ *  flags                   : 16
+ *  md_map                  : 64
+ *  mem_type                :  8
+ *
+ * exported_memh :
+ * [
+ *  address          : 64
+ *  length           : 64
+ *  ucp_context_uuid : 64
+ *  registration_id  : 64
+ * ]
+ *
+ * ucs_for_each_bit(md_map) {
+ *     tl_mkey_data_size_part1 :  8
+ *    [tl_mkey_data_size_part2 : 16 - packed only if tl_mkey_data_size > 255,
+ *                                    tl_mkey_data_size_part1 is set to 0]
+ *     tl_mkey_size            :  8
+ *     tl_mkey_packed          : <size>
+ *
+ *     exported_memh :
+ *     [
+ *      global_md_id_size   :  8
+ *      global_md_id        : <size>
+ *     ]
+ * }
+ */
+static ssize_t
+ucp_memh_exported_pack(const ucp_mem_h memh, void *buffer)
+{
+    ucp_context_h context            = memh->context;
+    uint64_t address                 = (uint64_t)ucp_memh_address(memh);
+    uint64_t length                  = ucp_memh_length(memh);
+    uct_mem_h *uct_memhs             = memh->uct;
+    void *p                          = buffer;
+    ucp_tl_md_t *tl_mds              = context->tl_mds;
+    uct_md_mkey_pack_params_t params = {
+        .field_mask = UCT_MD_MKEY_PACK_FIELD_FLAGS,
+        .flags      = UCT_MD_MKEY_PACK_FLAG_EXPORT
+    };
+    size_t memh_info_size            = ucp_memh_exported_info_packed_size();
+    size_t UCS_V_UNUSED memh_info_packed_size;
+    uct_md_attr_v2_t *md_attr;
+    char UCS_V_UNUSED buf[128];
+    ucs_status_t status;
+    ssize_t result;
+    ucp_md_index_t md_index;
+    size_t tl_mkey_size, global_id_size;
+    void *tl_mkey_buf, *global_id_buf;
+    size_t tl_mkey_data_size;
+
+    ucs_log_indent(1);
+
+    ucp_memh_common_pack(memh, &p, UCP_MEMH_BUFFER_FLAG_EXPORTED,
+                         memh_info_size);
+
+    *ucs_serialize_next(&p, uint64_t) = address;
+    *ucs_serialize_next(&p, uint64_t) = length;
+    *ucs_serialize_next(&p, uint64_t) = context->uuid;
+    *ucs_serialize_next(&p, uint64_t) = memh->reg_id;
+
+    memh_info_packed_size = UCS_PTR_BYTE_DIFF(buffer, p);
+    ucs_assertv(memh_info_size == memh_info_packed_size,
+                "memh_info_size %zu memh_info_packed_size %zu",
+                memh_info_size, memh_info_packed_size);
+
+    ucs_for_each_bit(md_index, context->export_md_map[memh->mem_type]) {
+        md_attr           = &tl_mds[md_index].attr;
+        tl_mkey_data_size = ucp_memh_exported_tl_mkey_packed_size(context,
+                                                                  md_index);
+        ucp_memh_info_size_pack(&p, tl_mkey_data_size);
+
+        tl_mkey_size = md_attr->exported_mkey_packed_size;
+        ucs_assertv((tl_mkey_size <= UINT8_MAX) && (tl_mkey_size != 0),
+                    "tl_mkey_size %zu", tl_mkey_size);
+        *ucs_serialize_next(&p, uint8_t) = tl_mkey_size;
+
+        tl_mkey_buf = ucs_serialize_next_raw(&p, void, tl_mkey_size);
```

Comment:
maybe can remove the local var tl_mkey_buf and pass ucs_serialize_next_raw directly to uct_md_mkey_pack_v2?

Reply:
done